tokio/src/fs/read_uring.rs

```diff
@@ -0,0 +1,195 @@
+use crate::fs::OpenOptions;
+use crate::runtime::driver::op::Op;
+
+use std::io::ErrorKind;
+use std::os::fd::OwnedFd;
+use std::path::Path;
+use std::{cmp, io};
+
+// this algorithm is inspired from rust std lib version 1.90.0
+// https://doc.rust-lang.org/1.90.0/src/std/io/mod.rs.html#409
+const PROBE_SIZE: usize = 32;
+const PROBE_SIZE_U32: u32 = PROBE_SIZE as u32;
+
+// Max bytes we can read using io uring submission at a time
+// SAFETY: cannot be higher than u32::MAX for safe cast
+const MAX_READ_SIZE: usize = u32::MAX as usize;
+
+pub(crate) async fn read_uring(path: &Path) -> io::Result<Vec<u8>> {
+    let file = OpenOptions::new().read(true).open(path).await?;
+
+    // TODO: use io uring in the future to obtain metadata
+    let size_hint: Option<usize> = file
+        .metadata()
+        .await
+        .map(|m| usize::try_from(m.len()).unwrap_or(usize::MAX))
+        .ok();
+
+    let fd: OwnedFd = file
+        .try_into_std()
+        .expect("unexpected in-flight operation detected")
+        .into();
+
+    let buf = Vec::with_capacity(size_hint.unwrap_or(0) + 1);
```

Comment:
We need a short comment to explain the reason of "plus one".

Reply:
let me know if its descriptive enough @ADD-SP